### ISSUE TITLE
wasm: bare `mut List` parameter — reallocating push not written back to caller (record-field half fixed in v0.27.9)

### ISSUE THREAD

**Issue by O6lvl4** [almide/almide#705]
## Summary

A `mut` parameter of a **List** type (or a `mut` record param with a `List` field) mutated in place via `list.push` works correctly on the **native** target but **silently diverges on the wasm target**: the mutation does not persist back to the caller, and a reallocating push corrupts the list.

This is the wasm-side counterpart of #703 (which fixed the native codegen + checker). #703's native fix is complete; this issue tracks the remaining wasm gap.

## Repro

```almide
fn push9(mut v: List[Int], x: Int) -> Int = { list.push(v, x); list.len(v) - 1 }
effect fn main() -> () = {
  var v: List[Int] = [10]
  let i = push9(v, 7)
  let j = push9(v, 99)
  println("len=" + int.to_string(list.len(v)) + " v0=" + int.to_string(v[0]))
}
```

```
native: len=3 v0=10            ✓ correct
wasm:   Error: index out of bounds   ✗
```

Starting from an empty list the divergence is silent rather than a trap:

```almide
// var v: List[Int] = []; push9(v,7); push9(v,99)
// native: len=2   wasm: len=0   (each call pushes into a discarded copy)
```

The same applies to the #703 record-field case (`mut b: Box` + `list.push(b.xs, v)`), and to slabhra's tape: `examples/mlp/mlp.almd` prints `loss = 1.020778409195663` on native but `loss = 0.0` on wasm.

## Root cause

`mut` params are passed by value on wasm (a pointer copy). A **List**'s runtime repr is inline `[len][data...]` at a single pointer, so a capacity-growing `list.push` **reallocates the whole header** → a new pointer the callee cannot write back to the caller's slot. The caller keeps pointing at the old (or empty) buffer.

`mut Map` works (see `spec/wasm_cross/mut_map_param.almd`, C-061) precisely because a Map is a *stable handle*: the top-level struct pointer never moves; `map.insert` mutates `{buckets_ptr, len}` in place, so a by-value pointer copy still observes the change.

## Fix directions

1. **Stable list handle** — represent lists as `[len][cap][data_ptr]` so a realloc only rewrites `data_ptr` inside a fixed-address header (matches the Map model). Touches the whole wasm list runtime.
2. **By-reference `mut` ABI** — pass `mut` params as a pointer-to-slot and write the realloc'd pointer back. ABI change to wasm call emission.

## Status

- Native fix: #703 (done). Native behavior locked by `spec/lang/mut_record_field_test.almd` (`// wasm:skip`, runs native).
- Downstream: slabhra's autograd tape is O(1) in-place on native; its wasm path waits on this issue.

**Comment by O6lvl4** [comment]
## Mostly fixed — slabhra now runs bit-identically on wasm

Three wasm codegen bugs fixed (branch `fable-llm/rccow-matrix-bytes`):

1. **Record-field push writeback** (`1598e066`). `emit_mutator_writeback` only handled a `Var` target; a `Member` target (`list.push(b.xs, v)` on a `mut b`) silently fell through, so a reallocating push dropped the new buffer. Now it stores the realloc'd pointer into the field slot `[base+offset]`. The record block is shared (passed by pointer on wasm), so it persists.

2. **Iteration-arena reclaim of an outer list grown by a stdlib mutator** (`14e16679`). The per-iteration heap arena (iter_scope) reclaimed a buffer that `list.push(out, …)` grew into an outer-scope var, when the loop body also allocated an inner temporary (e.g. `list.get(a, i) ?? d`). `expr_writes_outer_heap` now recognizes the mutating call. This was the real blocker — it made every numeric kernel (k_matmul, …) wrong/trapping on wasm, independent of the tape work.

3. **Same escape transitively through a user `mut`-param call** (`19525810`). `while … { op.add(t, …) }` on a `mut t: Tape` reallocs a field of `t` inside the callee; the buffer escapes via `t`. Collected per-function `mut`-param signatures (`emitter.mut_param_sigs`) and made the escape check consult them — precise, so read-only-helper loops keep the arena.

Result: all of slabhra (github.com/almide-graphics/slabhra) is byte-identical on native and wasm — MLP forward/backward, training (0.447 → ~3e-7), and the DLGN XOR gate network. Regression tests: `spec/lang/loop_arena_escape_test.almd`, `spec/lang/mut_record_field_test.almd` (both run cross-target).

## Remaining

The **bare `mut List` parameter** case is still wasm-divergent:

```almide
fn push9(mut v: List[Int], x: Int) -> Int = { list.push(v, x); list.len(v) - 1 }
// var v = [10]; push9(v, 20); push9(v, 30)  →  native len=3, wasm len=1
```

Unlike a record field (an addressable slot in a heap block), a bare list param's pointer lives in a non-addressable wasm **local**, so a reallocating push can't write the new pointer back to the caller's slot. This needs the by-reference `mut` ABI (spill the arg to a memory cell around the call). Tracked here; `spec/lang/mut_list_param_test.almd` carries `// wasm:skip` for it. slabhra does not hit this (its tape is a record).

**Comment by O6lvl4** [comment]
Update — the **record-field** half of this is FIXED and shipping in **v0.27.9** (PRs #710 native, #711 wasm, on develop → release PR #713): `list.push(b.xs, v)` on a `mut b` is now bit-identical on native and wasm (Member writeback + the loop/fn-scope reclamation escape scan extended to Member-rooted and transitive `mut`-param mutations). slabhra's autograd tape runs byte-identical native⇄wasm.

**Remaining (keeping this open):** the bare `mut List` **parameter** case — `fn push9(mut v: List[Int], x) = { list.push(v, x); … }` — still diverges on wasm (native len 3, wasm len 1). Unlike a record field (an addressable slot in a heap block), a bare list param's pointer lives in a non-addressable wasm **local**, so a reallocating push can't write the new pointer back to the caller. This needs the by-reference `mut` ABI (spill the arg to a memory cell around the call). `spec/lang/mut_list_param_test.almd` carries `// wasm:skip` for it.

(Separately, #712 tracks a `local var`'s record field loop-push dropping elements on both targets — a lowering bug #703 made reachable.)

**Comment by O6lvl4** [comment]
Scoping update: the **record-field** half of this shipped in **v0.27.9** (`list.push(b.xs, v)` on a `mut b` is now bit-identical native⇄wasm). The only case still open — and what this issue now tracks — is the **bare `mut List` parameter**: `fn push9(mut v: List[Int], x) = { list.push(v, x); … }` diverges on wasm (native len 3, wasm len 1), because a bare list param's pointer lives in a non-addressable wasm local, so a reallocating push can't write the new pointer back to the caller. Needs the by-reference `mut` ABI (spill the arg to a memory cell around the call). `spec/lang/mut_list_param_test.almd` carries `// wasm:skip`.

**Comment by O6lvl4** [comment]
Fixed on develop (`7aecd207`), CI green, ships in **v0.28.6** (PR #754).

## Root cause

`MutParamLoweringPass` (the wasm writeback transform) only handled **Unit-returning** callees in **statement position**. Your `fn push9(mut v: List[Int], x: Int) -> Int` in `let i = push9(v, 7)` matched neither: the callee body was never rewritten and the call site never wrote back, so the reallocating push landed in a discarded copy (`len=1` wasm vs `len=3` native; `len=0` growing from empty — the mlp loss=0.0).

## Fix

A bottom-up `IrMutVisitor` now rewrites **every** call position uniformly — statement, Bind/Assign RHS, nested expressions (`100 + push9(v, 2)`), and loop bodies (previously missed even for the Unit case). The call becomes

```
{ let (__mp_res, __mp_buf) = <call>; <writeback>; __mp_res }
```

- Value-returning callees return an `(orig, buffer)` tuple, destructured on the proven `BindDestructure` path (a hand-built TupleIndex read left the caller's list dangling when the tuple temp dropped).
- The writeback targets the argument **place**: a bare var assigns, a record-field arg (`push9(b.items, 7)`) FieldAssigns, a temp (no named place) skips — native mutates an unobservable temp there too.
- Bare-name collisions are excluded wholesale (the #692 class): with a changed signature, rewriting the callee but not every caller is an invalid module.
- Non-Unit **effect** fns are skipped for now (Result-wrap interplay) — they keep their previous behavior.

## Verified

The whole class, not just List: **List / String / Bytes × all call positions × var and field args** byte-match native. Contract **C-132** (ALS-M13 — spec text updated to name reallocating containers and field args) + fixture `spec/wasm_cross/mut_heap_param.almd` (13 assertions). Full bar green: `almide test` 280/280 (fallback baseline 13 regression-checked), wasm_cross release gate, workspace, frees-churn.
